config-proto-converter/src/main/java/org/hypertrace/config/proto/converter/ConfigProtoConverter.java

```diff
@@ -0,0 +1,24 @@
+package org.hypertrace.config.proto.converter;
+
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Message;
+import com.google.protobuf.MessageOrBuilder;
+import com.google.protobuf.Value;
+import com.google.protobuf.util.JsonFormat;
+
+public class ConfigProtoConverter {
+  private static final JsonFormat.Printer JSON_PRINTER = JsonFormat.printer();
+  private static final JsonFormat.Parser JSON_PARSER = JsonFormat.parser().ignoringUnknownFields();
+
+  public static void mergeFromValue(Value value, Message.Builder builder)
+      throws InvalidProtocolBufferException {
+    JSON_PARSER.merge(JSON_PRINTER.print(value), builder);
```

Comment:
I think this would throw an error in the case where `value` indicates an empty/null value (`Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build()`). Need to take care of that.

Might as well be helpful if we have a test for this.

Reply:
I think this is debatable when looking at the generic case which this converter captures. If my schema'd message has a valid config using only its default values, such behavior would break those conversions. 

For example, imagine my message had a single boolean field, enabled/disabled:
```
message MyConfig {
    bool enabled = 1;
```

The disabled case would serialize as an empty message due to how proto handles default values, and it would deserialize back to us as an empty value. Currently, this converter would then correctly return an instance of `MyConfig` which would signify disabled, but the proposed change would cause it to error.

I'd prefer to leave validation of the resulting message to the caller.

Reply:
Actually, empty and null value are different here(I confused the two in my previous comment. Sorry for that). I agree with you in the case of empty value.

But now consider this case of null value - we make a `getConfig` call to the generic config service for a config resource for which no config is defined yet. In such case, generic config service returns a null value (actually a representation for null) - `Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build()`. Now if we pass this value to this method with the appropriate builder, it will throw an error as the `JSON_PRINTER` will print `null` and `JSON_PARSER` will be unable to parse it.

Reply:
Gotcha - you're talking about the case where the value returned is a top level null. 

First, what are the scenarios where this is returned? Any value that had been written would not be a top level null. Is it only when I ask for a config by context(s) and it doesn't exist (nor a default)? In that case, should the service itself throw an error or return a null? I could be convinced either way there, but I'd probably lean towards throwing as that's generally what API semantics indicate (you'd throw a 404 in the case of REST).

Second, assuming we are returning null in that case, what would we expect this converter do? The caller is expecting a message back in their schema - so the only options I can think of (feel free to add alternate ideas) are:
- Throw: this is what it would currently do as you mentioned `Unexpected exception thrown: java.lang.RuntimeException: com.google.protobuf.InvalidProtocolBufferException: Expect message object but got: null`
- Optional: I think returning an optional for some forms of invalid data but throwing for others is confusing. so this means we'd have to swallow and log any conversion errors and return an empty optional. The caller would likely end up just throwing again anyway (since this is presumably a get call that can't return), this time without the context of the error.
- DefaultInstance: This feels wrong for the same reasons discussed above around empty message - an empty message may be valid and now be indistinguishable from an invalid null.

Reply:
> Is it only when I ask for a config by context(s) and it doesn't exist (nor a default)? 

Yes, that's the only case. 

> In that case, should the service itself throw an error or return a null?

I think this depends on the service. As you pointed, the service might want to throw an error. But this may not be the case always. For instance, the service may want to get config for a config resource, do something with it based on the request(it may want to have its own merging logic) and then upsert it. So, the first time it tries to get a config for a config resource, it will get a null value.

> Second, assuming we are returning null in that case, what would we expect this converter do?

I think returning an `Optional` would be fine. We just check if `value` is null(i.e. `value.equals(Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build())`), then return an empty optional. We don't have to swallow any conversion errors here, right?

Reply:
> I think this depends on the service. As you pointed, the service might want to throw an error. But this may not be the case always. For instance, the service may want to get config for a config resource, do something with it based on the request(it may want to have its own merging logic) and then upsert it. So, the first time it tries to get a config for a config resource, it will get a null value.

It's certainly true that we can make either work. The service can throw, the caller can catch it, recover and proceed - or the service can return null, the caller can check it and throw. So it's more about the api design rather than what use cases does each enable. My preference is that the api exposes one error path rather than two separate ones.

> I think returning an Optional would be fine. We just check if value is null(i.e. value.equals(Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build())), then return an empty optional. We don't have to swallow any conversion errors here, right?

I think this directly follows from the prior point. If we maintain the response/null/error behavior of the service, the optional would allow us to mimic it and probably makes sense. If we think that behavior should coalesce to just response/error, this should too.

Reply:
@GurtejSohi - I do think the converter should match the behavior of the api, but right now they're inconsistent. Do we think it makes sense to make the behavior change and throw instead of returning null (in which case, no change required here) or to leave the behavior as is and update the converter?

Reply:
> I do think the converter should match the behavior of the api, but right now they're inconsistent.

Oh, right. For a second, I thought the generic config service was in another repo :)

> Do we think it makes sense to make the behavior change and throw instead of returning null (in which case, no change required here) or to leave the behavior as is and update the converter?

My preference is that we maintain response/null/error behaviour (and update the converter) because I think it helps the consumer to distinguish between "config not found" and "server error" without having to check/process the error message.
But, I'm fine with the other approach as well. Please feel free to take the call.

Reply:
I'm going to take it in a separate PR, but unless anyone has strong feelings (@buchi-busireddy  ?), I'm going to change the behavior of the API to return a status `5` https://grpc.github.io/grpc/core/md_doc_statuscodes.html (grpc equivalent to 404) if the response is not found. My reasoning is that, in the not found case we're going to have to parse something - either the error or the response, to disambiguate it from what we're calling a "real" error or "valid" response. I'd rather use the GRPC standard error code for this case rather than building an api-specific body that indicates the same. The client utility I see as completely orthogonal.